src/bitcoin-cli.cpp

```diff
@@ -340,13 +340,20 @@ int CommandLineRPC(int argc, char *argv[])
         if (args.size() < 1)
             throw std::runtime_error("too few parameters (need at least command)");
         std::string strMethod = args[0];
-        UniValue params = RPCConvertValues(strMethod, std::vector<std::string>(args.begin()+1, args.end()));
+        auto params =
+            RPCConvertValues(strMethod, std::vector<std::string>(args.begin()+1, args.end()));
+
+        if (!params.has_value()) {
+            params.map_error([&](auto failure) {
+                throw std::runtime_error(FormatConversionFailure(strMethod, failure));
+            });
+        }
 
         // Execute and handle connection failures with -rpcwait
         const bool fWait = GetBoolArg("-rpcwait", false);
         do {
             try {
-                const UniValue reply = CallRPC(strMethod, params);
+                const UniValue reply = CallRPC(strMethod, params.value());
```

Comment:
I'm not sure how much more invasive it would be, but what about, instead of throwing on line 348, we instead invoke `CallRPC` in whatever way is required to generate the help text, and then glue that together with the format conversion failure so that you get both "here's what you did wrong" and "here's what you should do"?

Reply:
Yeah – previously failing on the client side only errored, without much info, but now we error more on the client side, which means we lose help in some cases. I think we can always make a `help` RPC request in that case, either bare (in the case of no command or undefined command being sent) or for the specific command in the case of wrong number of params or format conversion failure.

In future, I feel like the help definitions should be pulled out into common.cpp/.h so even those messages can be provided without making an actual request (again, something that would be even easier if we had an actual schema … maybe I should put up a draft with my partial RPC schema from when I started, which could also be an alternative approach to support keyed arguments).